state/factory/statedb.go

```diff
@@ -446,27 +449,27 @@ func (sdb *stateDB) DeleteTipBlock(_ *block.Block) error {
 
 // State returns a confirmed state in the state factory
 func (sdb *stateDB) State(s interface{}, opts ...protocol.StateOption) (uint64, error) {
-	sdb.mutex.Lock()
-	defer sdb.mutex.Unlock()
-
 	cfg, err := processOptions(opts...)
 	if err != nil {
 		return 0, err
 	}
-
-	return sdb.currentChainHeight, sdb.state(cfg.Namespace, cfg.Key, s)
+	sdb.mutex.RLock()
+	h := sdb.currentChainHeight
+	sdb.mutex.RUnlock()
+	return h, sdb.state(cfg.Namespace, cfg.Key, s)
```

Comment:
state will read data from dao and then deserialize it. If we don't lock the whole struct, will `h` be inconsistent with the result read from `sdb.state`?

Reply:
fixed